### ISSUE TITLE
configure assumes ldconfig can be called and that CuDNN is installed in a system library location

### ISSUE THREAD

**Issue by jowagner** [tensorflow/tensorflow#6762]
The configure script calls "ldconfig" to locate the CuDNN library but this makes 2 assumptions that can be wrong: 
1. It assumes that "ldconfig" can be called. However, on some GNU/Linux systems such as openSUSE Leap 42.1, only root has it in their path. (On Leap 42.1, normal users get the desired output from "/sbin/ldconfig -p".)
2. It assumes that the CuDNN library is installed in a system library location that is searched by ldconfig. However, users may install it in their home directory and use LD_LIBRARY_PATH to load it for various reasons. I, for example, was not fully satisfied from reading the licence terms that I am allowed to share the library with other users of the system and therefore did not install it system-wide as root.

Severity is low as normally the CuDNN library location should be auto-detected from the CuDNN library path before "ldconfig" is tried.

How to reproduce:
1. git clone this repo on a system with ldconfig not in your PATH
2. cd tensorflow and run ./configure
3. when asked for the CuDNN path, enter a valid and readable path (to avoid other errors), for example /usr/bin, that does not contain CuDNN
4. Error message: ./configure: line 337: ldconfig: command not found

Workaround options:
- First double check the CuDNN library path (the default path suggests that the last component "lib64" or "lib" should be omitted) and verify that standard folder names are used under this location for library and header files, for example "lib64" and "include".
- If you can call /sbin/ldconfig and if CuDNN is in a standard system library location the user can temporarily add /sbin to their PATH before running "./configure"
- Edit the "configure" script to provide the location manually, for example CUDNN_PATH_FROM_LDCONFIG="$HOME/local/cuda/lib64/libcudnn.so"

A fix should test whether "ldconfig" can be called before trying to call it. Also try /sbin/ldocnfig. Given that the library should normally be found already earlier in the code using the user-provided CuDNN path, a simple solution may be to set CUDNN_PATH_FROM_LDCONFIG to the empty string if ldconfig cannot be called.

If you like to automate the locating of the library as much as possible, you could furthermore check all locations in LD_LIBRARY_PATH.

**Comment by jart** [comment]
Thanks for bringing this to our attention. It would be nice if the configure script was smarter about discovering this library. As for ldconfig, wouldn't it work if it was just changed to `/sbin/ldconfig`? Where else could it be?

Please be advised that OpenSUSE isn't in our support matrix. Our goal is to support Ubuntu Linux 14 LTE to 16. Mac OS X El Capitan. CentOS 7 and above. So it might be best if we marked this contributions welcome.

**Comment by jowagner** [comment]
TLDR: A solution may be to set PATH="$PATH":/sbin at the start of the configure script. This way, if the system or the user has a preferred "ldconfig" somewhere else it will be used and if not the script will automatically fall back to /sbin/ldconfig.

The FHS http://www.pathname.com/fhs/pub/fhs-2.3.html only mentions "ldconfig" in the /sbin section. From this point of view, it should be safe to assume that "ldconfig" is always in /sbin (if it is installed at all -- the FHS says it's optional). However, it's best practise to respect the preferences expressed in PATH.

My first thought was that "ldconfig" cannot be in /sbin on the supported platforms as otherwise the "configure" script would not work on these platforms because normal users don't have /sbin in their PATHs. However, most Ubuntu users have sudoer rights. That's probably why this issue hasn't been noticed before.

**Comment by jart** [comment]
That makes sense. In that case, we'd be grateful if you joined the ranks of TensorFlow contributors and sent us the one line change you proposed. CC me and I'll review it.

**Comment by jowagner** [comment]
Trivial code change #6848 rejected as drpngx insists on Google's Contributor License Agreement to be signed and I don't have time to review it and the documents it refers to.

Anybody with CLA please take the 30 seconds to implement "temporarily add /sbin to PATH for ldconfig call" in your own way and contribute a patch (total time <5 minutes). (Don't copy my solution. Even though I would be ok with you copying it as I think it is trivial code, a Google bot may detect that it is identical or similar to a previously rejected pull request and reject it on these grounds.)

**Comment by jart** [comment]
drpngx was only following Google policy. The policy is designed to protect the interests of the TensorFlow community. It's applied evenly and uniformly to all contributions. So I hope you'll reconsider. I could easily make the change myself. I just feel that since you were able to identify an issue with TensorFlow along with its solution, you deserve to be able to take public credit as the person who solved it.
